src/core/thread/mesh_forwarder.cpp

```diff
@@ -206,6 +214,168 @@ void MeshForwarder::ScheduleTransmissionTask()
+            mSrcMatchEnabled = false;
+        }
+        else
+        {
+            // succeed in adding to source match table
+            aChild.mAddSrcMatchEntryPending = false;
+        }
+    }
+
+exit:
+
+    if (!mSrcMatchEnabled)
+    {
+        if (aChild.mAddSrcMatchEntryPending && (aChild.mQueuedIndirectMessageCnt == 1))
+        {
+            if (aChild.mAddSrcMatchEntryShort)
+            {
+                mSrcMatchOverflowShortCnt++;
+            }
+            else
+            {
+                mSrcMatchOverflowExtCnt++;
+            }
+
+        }
+
+        otLogDebgMac("Overflow(Short,Ext)=(%d, %d), Released(Short, Ext)=(%d,%d)\n",
+                     mSrcMatchOverflowShortCnt,
+                     mSrcMatchOverflowExtCnt,
+                     mSrcMatchReleasedShortCnt,
+                     mSrcMatchReleasedExtCnt);
+    }
+
+    return;
+}
+
+void MeshForwarder::ClearSrcMatchEntry(Child &aChild)
+{
+    Mac::Address macAddr;
+
+    if (aChild.mAddSrcMatchEntryShort)
+    {
+        macAddr.mLength = sizeof(macAddr.mShortAddress);
+        macAddr.mShortAddress = aChild.mValid.mRloc16;
+    }
+    else
+    {
+        macAddr.mLength = sizeof(macAddr.mExtAddress);
+        memcpy(macAddr.mExtAddress.m8, aChild.mMacAddr.m8, sizeof(macAddr.mExtAddress));
+    }
+
+    if (mMac.ClearSrcMatchEntry(macAddr) == kThreadError_None)
+    {
+        // if waiting for more source table resources
+        if (!mSrcMatchEnabled)
+        {
+            if (aChild.mAddSrcMatchEntryShort)
+            {
+                mSrcMatchReleasedShortCnt++;
+            }
+            else
+            {
+                mSrcMatchReleasedExtCnt++;
+            }
+        }
+    }
+    else
+    {
+        // if finished queued messages for SED which is not added into the source match table
+        if (aChild.mAddSrcMatchEntryShort)
+        {
+            mSrcMatchOverflowShortCnt--;
+        }
+        else
+        {
+            mSrcMatchOverflowExtCnt--;
+        }
+
+        aChild.mAddSrcMatchEntryPending = false;
+    }
+
+    otLogDebgMac("Overflow(Short,Ext)=(%d, %d), Released(Short, Ext)=(%d,%d)\n",
+                 mSrcMatchOverflowShortCnt,
+                 mSrcMatchOverflowExtCnt,
+                 mSrcMatchReleasedShortCnt,
+                 mSrcMatchReleasedExtCnt);
+
+    // use tasklet for re-enable source match to avoid possible disorder due to
+    // possible repeatedly disable/enable when many SEDs exists
+    mUpdateSrcMatchTask.Post();
+}
+
+void MeshForwarder::HandleUpdateSrcMatchTask(void *aContext)
+{
+    MeshForwarder *obj = reinterpret_cast<MeshForwarder *>(aContext);
+    obj->HandleUpdateSrcMatchTask();
+}
+
+void MeshForwarder::HandleUpdateSrcMatchTask()
+{
+    // if total available source match table resources more than total required resources
+    if (!mSrcMatchEnabled &&
+        (mSrcMatchReleasedShortCnt + 2 * mSrcMatchReleasedExtCnt) >= (mSrcMatchOverflowShortCnt + 2 * mSrcMatchOverflowExtCnt))
```

Comment:
The overflow logic here is to take full use of source match table which is shared by short and extended address in CC2538.  But you are right, they may be separate for other radios.

Should we decide the overflow logic by adding an interface to indicate whether source match table for short/extended address separate, or use your above suggestions for now?